hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrngSystemPrecompiledContract.java

```diff
@@ -0,0 +1,283 @@
+	public static final String PRNG_PRECOMPILE_ADDRESS = "0x169";
+	private final PrngLogic prngLogic;
+	private final SideEffectsTracker tracker;
+	private final EntityCreator creator;
+	private HederaStackedWorldStateUpdater updater;
+	private final RecordsHistorian recordsHistorian;
+	private final PrecompilePricingUtils pricingUtils;
+	private final Supplier<Instant> consensusNow;
+	private final LivePricesSource livePricesSource;
+	private final GlobalDynamicProperties properties;
+	private long gasRequirement;
+
+	@Inject
+	public PrngSystemPrecompiledContract(
+			final GasCalculator gasCalculator,
+			final PrngLogic prngLogic,
+			final ExpiringCreations creator,
+			final SideEffectsTracker tracker,
+			final RecordsHistorian recordsHistorian,
+			final PrecompilePricingUtils pricingUtils,
+			final Supplier<Instant> consensusNow,
+			final LivePricesSource livePricesSource,
+			final GlobalDynamicProperties properties) {
+		super(PRECOMPILE_NAME, gasCalculator);
+		this.prngLogic = prngLogic;
+		this.creator = creator;
+		this.tracker = tracker;
+		this.recordsHistorian = recordsHistorian;
+		this.pricingUtils = pricingUtils;
+		this.consensusNow = consensusNow;
+		this.livePricesSource = livePricesSource;
+		this.properties = properties;
+	}
+
+	@Override
+	public long gasRequirement(final Bytes bytes) {
+		return gasRequirement;
+	}
+
+	@Override
+	public PrecompileContractResult computePrecompile(final Bytes input, final MessageFrame frame) {
+		ExpirableTxnRecord.Builder childRecord;
+		Bytes result;
+		Bytes randomNum = Bytes.EMPTY;
+
+		try {
+			gasRequirement = calculateGas();
+			validateTrue(frame.getRemainingGas() >= gasRequirement, INSUFFICIENT_GAS);
+
+			final var selector = input.getInt(0);
+			randomNum = generatePseudoRandomData(selector, input);
+
+			childRecord = createSuccessfulChildRecord(randomNum, frame, input);
+			result = SUCCESS_RESULT;
+
+		} catch (InvalidTransactionException e) {
+			childRecord = createUnsuccessfulChildRecord(e.getResponseCode(), frame,
+					e.isReverting() ? e.getRevertReason() : Bytes.EMPTY);
+			result = resultFrom(e.getResponseCode());
+		} catch (Exception e) {
+			log.warn("Internal precompile failure", e);
+			childRecord = createUnsuccessfulChildRecord(FAIL_INVALID, frame, Bytes.EMPTY);
+			result = resultFrom(FAIL_INVALID);
+		}
+
+		final var parentUpdater = updater.parentUpdater();
+		if (parentUpdater.isPresent()) {
+			final var parent = (AbstractLedgerWorldUpdater) parentUpdater.get();
+			parent.manageInProgressRecord(recordsHistorian, childRecord, body(randomNum, input));
+		} else {
+			throw new InvalidTransactionException("PRNG precompile frame had no parent updater", FAIL_INVALID);
+		}
+
+		return result == null ?
+				PrecompiledContract.PrecompileContractResult.halt(null, Optional.of(ExceptionalHaltReason.NONE))
+				: PrecompiledContract.PrecompileContractResult.success(result);
+	}
+
+	Bytes generatePseudoRandomData(final int selector, final Bytes input) {
```

Comment:
Could this be improved as below since you're already passing in `input`?
```suggestion
	Bytes generatePseudoRandomData(final Bytes input) {
                final var selector = input.getInt(0);
```

Reply:
Fixed. Thanks !